Improve VAO core profile performance

unknownbrackets opened the pull request:
Based on @Sonicadvance1's suggestions, I tried using a buffer cache based on the size of the buffer (so they would not get resized at runtime, if possible.)  I also increased the buffer cache size.

This helped significantly.  For me, performance is now the same/better with `gl_extensions.IsCoreContext = true;` compared to with it off.  Before, it was significantly worse.

This may also help even with it off, because it will prefer a buffer that has the same size for the vertex cache as well.

-[Unknown]

unknownbrackets committed:
Try to add support for GL_KHR_debug.

For some reason I'm still not seeing this work in a core profile...

unknownbrackets committed:
Reuse buffers of the same size for dynamic data.

This is a lot faster than changing their sizes all the time.

unknownbrackets committed:
thin3d: Use GL_COLOR_LOGIC_OP.

daniel229 commented:
Breaks Gundam VS Gundam NEXTPLUS
![02](https://cloud.githubusercontent.com/assets/3481559/11973259/b3099d64-a992-11e5-9e3c-2a3f91724ff0.png)

unknownbrackets commented:
Does it help to change VERTEXCACHE_NAME_CACHE_FULL_SIZE back to 80 at the top?

-[Unknown]

daniel229 commented:
No,it doesn't.

When it's corrupted,there is a log like this
```
52:22:206 user_main    W[G3D]: GLES\TextureCache.cpp:1997 Texture cache ran out of GPU memory; switching to low memory mode
```

unknownbrackets commented:
Oh, sounds like I introduced a leak somewhere...

-[Unknown]

unknownbrackets committed:
Decimate buffer objects less often.

I think it was previously deleting used objects, oops.

unknownbrackets committed:
Display types at different log levels.

Just to make errors stand out more, mainly.

unknownbrackets commented:
Does this resolve that at all?

-[Unknown]

daniel229 commented:
Yeah,looks fine now.

daniel229 commented:
Split/Second getting more stutterings.

unknownbrackets committed:
Decimate buffer names less frequently.

Sometimes there can be a lot, so let's decimate in chunks, and only based
on memory pressure.

unknownbrackets commented:
Hmm.  Just to make sure it's not decimation, can you put `double st = real_time_now();` right before `void TransformDrawEngine::DecimateBuffers() {`, and then `NOTICE_LOG(HLE, "Decimating buffers for %fms", (real_time_now() - st) * 1000.0f);` at the end of the function?

For me, 3000 buffers get decimated at times in Star Ocean (with VAO forced on), for example, which takes 5ms.  I've improved that.

If it's still an issue, can you try changing `#define VERTEXCACHE_NAME_CACHE_FULL_BYTES (1024 * 1024)` to like `#define VERTEXCACHE_NAME_CACHE_FULL_BYTES (1024 * 1024 * 10)`?

I'm actually surprised at how often we use >= 1MB in vertex buffers.  I wonder how much it would help (bandwidth wise) to decrease this...

-[Unknown]

unknownbrackets committed:
Use the new way for GL_EXTENSIONS in reporting too.

daniel229 commented:
latest commit still not help it.
Log here.
https://gist.github.com/daniel229/317a1e7c577ae8fa4704

<code>#define VERTEXCACHE_NAME_CACHE_FULL_BYTES (1024 * 1024 * 10)</code> not help it either.

unknownbrackets commented:
Woah: Decimating buffers for 35910.653328ms
Decimating buffers for 95689.248679ms

That's not good.  I guess they're almost definitely leaking.  Hmmm...

-[Unknown]

unknownbrackets committed:
Oops, actually free old vertex cache buffers.

And reuse them, since, why not?

unknownbrackets commented:
Okay, right, I was only focusing on buffers, I forgot to check with IsCoreContext off so I could see only the cached arrays.  They were leaking - but should be fine now.

-[Unknown]

daniel229 commented:
Great,fixed.

daniel229 commented:
Are thoes errors important?When I exit God of War Ghost Of Sparta
![01](https://cloud.githubusercontent.com/assets/3481559/11988215/2664a97c-aa27-11e5-8f58-25b48fd5b940.png)

unknownbrackets committed:
Prevent leaking VBOs on shutdown also.

unknownbrackets commented:
Ah, thanks for catching, yes those errors are important too.

-[Unknown]

unknownbrackets commented:
So I think this is all good now.  I guess we could enable VAOs by default now maybe, but it might be slower on other devices.

-[Unknown]
